beta-operator-rewards/rewards-requirements.ts

```diff
@@ -0,0 +1,478 @@
+    //    ],
+
+    peersData.push(peerData);
+  }
+
+  console.log("peersData: ", JSON.stringify(peersData, null, 2));
+
+  // TODO: calculate weighted authorization for a given staking provider
+  // <staking provider address>: {
+  //   beneficiary: <address>,
+  //   weightedAuthorization: <amount> // uptimeCoeficient * min(beaconAuthorization, tbtcAuthorization)
+  // }
+
+  // TODO: Save to file
+  // - all requirements
+  // - weighted authorization for NU team
+  // const jsonObject = await convertToJSON(peersData);
+  // fs.writeFileSync(peersDataFile, JSON.stringify(jsonObject, null, 2));
+}
+
+async function getAuthorization(
+  application: Contract,
+  intervalEvents: any[],
+  postEvents: any[],
+  stakingProvider: string,
+  startRewardsBlock: number,
+  endRewardsBlock: number
+) {
+  // When there were no events during the rewards interval, then we fetch the
+  // authorization after the interval which was the same as for the actual
+  // rewards interval.
+  if (intervalEvents.length == 0) {
+    // Events that were emitted between the [end:firstEventDate|currentDate] dates.
+    // This is used to fetch the authorization that was allocated during the rewards
+    // interval.
+    const postIntervalEvents = await filterEventsByStakingProvider(
+      postEvents,
+      stakingProvider
+    );
+
+    return await authorizationPostRewardsInterval(
+      postIntervalEvents,
+      application,
+      stakingProvider
+    );
+  }
+
+  // There is at least one event emitted during the rewards interval
+  const applicationEvents = intervalEvents.filter((obj) => {
+    return obj.args.application == application.address;
+  });
+
+  return await authorizationForRewardsInterval(
+    applicationEvents,
+    startRewardsBlock,
+    endRewardsBlock
+  );
+}
+
+async function filterEventsByApplications(events: any[]) {
+  const filteredEvents = new Array();
+  for (let j = 0; j < events.length; j++) {
+    const event = events[j];
+    if (
+      event.args.application == RandomBeaconAddress ||
+      event.args.application == WalletRegistryAddress
+    ) {
+      filteredEvents.push(event);
+    }
+  }
+
+  return filteredEvents;
+}
+
+async function filterEventsByStakingProvider(
+  events: any[],
+  stakingProvider: string
+) {
+  const filteredEvents = new Array();
+  for (let j = 0; j < events.length; j++) {
+    const event = events[j];
+    if (event.args.stakingProvider == stakingProvider) {
+      filteredEvents.push(event);
+    }
+  }
+
+  return filteredEvents;
+}
+
+// Calculates the weighted authorization for rewards interval based on events.
+// The general idea of weighted rewards calculation describes the following example.
+// Please note that this example operates on dates for simplicity purposes,
+// however the actual calculation is based on block numbers.
+// Ex.
+// events:         ^     ^      *    *   ^
+// timeline:  -|---|-----|------|----|---|--------|-->
+//         Sep 0   3     8      14  18   22       30
```

Comment:
this graph is awesome

Reply:
![image](https://user-images.githubusercontent.com/3156420/197889469-fffc31ee-1acc-4893-87d5-b8e9ea6ab7ec.png)